<font size="+3" color='#053c96'><h2><center> Classically Punk</h2></center></font>
<figure>
<center><img src ="https://images.unsplash.com/photo-1487180144351-b8472da7d491?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=872&q=80" width = "750" height = '600' alt="Classical Punk"/>
<font size="0" color='#053c96'><h4><center> Photo Credit: Unsplash</h4></center></font>

<font size="+2" color='#053c96'><b> Contributor</b></font>  
<font size="+0" ><b> Umar Kabir</b></font>  

<a id='table-of-contents'></a>
[Table of Contents](#table-of-contents)

- [Introduction](#introduction)
  * [Overview](#overview)
  * [Problem Statement](#problem-statement)
  * [Objectives](#goals)
- [Importing Libraries](#importing-dependencies)
- [Data](#data)
- [Exploratory Data Analysis](#exploratory-data-analysis)
  * [Data Exploration](#data-exploration)
  * [Data Visualization](#data-visualization)
  * [Summary Statistics](#summary-statistics)
  * [Feature Correlation](#feature-correlation)
- [Data Preparation](#data-preparation)
  * [Data Cleaning](#data-cleaning)
  * [Feature Engineering](#feature-engineering)
  * [Data Transformation](#data-transformation)
- [Modeling](#modeling)
  * [Model Selection](#model-selection)
  * [Model Training](#model-training)
  * [Model Evaluation](#model-evaluation)
  * [Hyperparameter Tuning](#hyperparameter-tuning)
- [Results](#results)
  * [Analysis Results](#analysis-results)
  * [Model Performance](#model-performance)
  * [Feature Importance](#feature-importance)
  * [Implications](#implications)
- [Conclusion](#conclusion)
  * [Summary](#summary)
  * [Limitations](#limitations)
  * [Recommendations](#recommendations)
- [References](#references)

<a id='introduction'></a>
<font size="+2" color='#053c96'><b> 1. Introduction</b></font>  
[back to top](#table-of-contents)  

Classically Punk is a music genre classification project that combines different elements of music to develop a machine learning model that predicts music genre.  

In this project, we aim to use machine learning techniques to automatically classify different musical genres, from audio snippets. To achieve this, we will need to find a library that can read music files and extract features from them, such as tempo, pitch, and melody. We will then use these features to train a machine learning model that can classify different genres of music.  

The deliverables for this project include a presentation with slides on how we classified the music, as well as assumptions, implications, and other important information, and code that the DevOps team can push to production.

<a id='overview'></a>
<font size="+1" color='#780404'><b> 1.1 Overview</b></font>  
[back to top](#table-of-contents)  

The project aims to develop a machine learning application that can automatically classify different musical genres from audio snippets. The main steps involved in the project include finding a library that can read music files and extract features from them, identifying relevant features for classification, and training a machine learning model to classify different genres of music. The project also involves handling large data sets and analyzing media files to generate data and identify patterns. 

<a id='problem-statement'></a>
<font size="+1" color='#780404'><b> 1.2 Problem Statement</b></font>  
[back to top](#table-of-contents)  

The problem statement for this project is the difficulty in manually classifying large collections of music into different genres. This process can be time-consuming and prone to errors, as it requires a deep understanding of the characteristics of each genre. Furthermore, as the amount of music available online continues to grow, it becomes increasingly challenging to keep up with the task of categorizing music by hand. The goal of this project is to develop a machine learning application that can automate the process of music classification, making it faster, more accurate, and scalable. The application will use features extracted from audio files to train a machine learning model that can classify different genres of music, including Classically Punk, without the need for human intervention.

<a id='goals'></a>
<font size="+1" color='#780404'><b> 1.3 Objectives</b></font>  
[back to top](#table-of-contents)  

1. To identify a suitable library for reading music files and extracting features from them.
2. To determine relevant features that can be used for music genre classification, such as tempo, pitch, and melody.
3. To preprocess the audio data, such as removing noise and converting it into a format suitable for machine learning algorithms.
4. To train a machine learning model on the audio data using a suitable algorithm, such as neural networks or decision trees.
5. To evaluate the performance of the machine learning model using appropriate metrics, such as accuracy, precision, and recall.
6. To optimize the model's performance by tuning hyperparameters and experimenting with different algorithms.
7. To develop a user-friendly interface for the application that allows users to upload audio files and receive genre classification results.
8. To present the results of the project, including the classification accuracy and the features that were most relevant for genre classification.
9. To deliver code that can be easily deployed by the DevOps team for use in a production environment.
10. - Speech activity detection: Speech activity detection is the task of identifying the segments of an audio signal that contain speech. This can be done by looking for changes in the energy of the signal, the zero crossing rate, or the pitch.
- Speaker identification: Speaker identification is the task of identifying the speaker of an audio signal. This can be done by looking for features that are unique to each speaker, such as the vocal tract shape or the way that they pronounce certain words.
- Music genre classification: Music genre classification is the task of classifying an audio signal into a particular genre of music. This can be done by looking for features that are commonly associated with different genres of music, such as the tempo, the pitch, or the rhythm.
- Sound event detection: Sound event detection is the task of identifying the different types of sounds that are present in an audio signal. This can be done by looking for changes in the energy of the signal, the spectral content of the signal, or the temporal structure of the signal.
- Automatic music transcription: Automatic music transcription is the task of converting an audio recording of music into a musical score. This can be done by looking for features that are associated with different musical notes, such as the pitch, the duration, and the timbre.

<a id='importing-dependencies'></a>
<font size="+2" color='#053c96'><b> 2. Importing Libraries</b></font>  
[back to top](#table-of-contents)

In [12]:
import sys
# Insert the parent path relative to this notebook so we can import from the src folder.
sys.path.insert(0, "..")

from src.dependencies import *
from src.functions import *

<a id='data'></a>
<font size="+2" color='#053c96'><b> 3. Data</b></font>  
[back to top](#table-of-contents)  

The dataset was used for the well-known paper in genre classification "Musical genre classification of audio signals" by G. Tzanetakis and P. Cook in IEEE Transactions on Audio and Speech Processing 2002.

In [2]:
df = pd.read_csv('../data/data.csv')

<a id='exploratory-data-analysis'></a>
<font size="+2" color='#053c96'><b> 4. Exploratory Data Anaysis</b></font>  
[back to top](#table-of-contents)

<a id='data-exploration'></a>
<font size="+1" color='#780404'><b> 4.1 Data Exploration</b></font>  
[back to top](#table-of-contents)

In [3]:
df.shape

(1000, 22)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 22 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Genre               1000 non-null   object 
 1   Filename            1000 non-null   object 
 2   Beats               1000 non-null   object 
 3   SR                  1000 non-null   int64  
 4   Central Moments     1000 non-null   object 
 5   Zero Crossing Rate  1000 non-null   object 
 6   RMSE                1000 non-null   object 
 7   Tempo               1000 non-null   float64
 8   Spectral Contrast   1000 non-null   object 
 9   Spectral Roll-off   1000 non-null   object 
 10  MFCC                1000 non-null   object 
 11  Chroma              1000 non-null   object 
 12  Spectral Centroid   1000 non-null   object 
 13  Spectral Bandwidth  1000 non-null   object 
 14  Spectral Flatness   1000 non-null   object 
 15  Loudness            1000 non-null   object 
 16  Complex

In [5]:
df.head()

Genre         Filename                                              Beats  \
0  blues  blues.00000.wav  [  11   32   53   73   94  114  135  156  177 ...   
1  blues  blues.00001.wav  [   3   41   80  118  156  194  231  267  304 ...   
2  blues  blues.00002.wav  [  13   30   46   64   80   96  112  129  146 ...   
3  blues  blues.00003.wav  [  28   69  113  155  195  236  279  320  360 ...   
4  blues  blues.00004.wav  [   8   27   48   67   85  104  125  145  164 ...   

      SR                                    Central Moments  \
0  22050  [ 1.9598199e-10  1.9793237e-02 -5.6590850e-04 ...   
1  22050  [2.6082897e-10 1.1581921e-02 8.0263242e-05 7.1...   
2  22050  [-1.2630751e-09  3.3572275e-02 -9.4693369e-04 ...   
3  22050  [3.1796636e-09 2.6253546e-02 5.5292981e-05 2.9...   
4  22050  [2.0570905e-10 1.0682435e-02 1.4974119e-04 7.9...   

                                  Zero Crossing Rate  \
0  [0.06494141 0.08300781 0.10791016 ... 0.086425...   
1  [0.015625   0.01806641 0.02148438 ... 0.070312...   
2  [0.04638672 0.06396484 0.0859375  ... 0.086914...   
3  [0.01025391 0.01464844 0.01855469 ... 0.034179...   
4  [0.03955078 0.06494141 0.08935547 ... 0.065917...   

                                                RMSE       Tempo  \
0  [0.01908039 0.03683019 0.0497628  ... 0.140552...  123.046875   
1  [0.04788139 0.10656705 0.14316952 ... 0.137602...   67.999589   
2  [0.12294307 0.14580017 0.15829462 ... 0.134463...  161.499023   
3  [0.10660144 0.13829172 0.16152988 ... 0.017918...   63.024009   
4  [0.05263326 0.07990757 0.09952808 ... 0.054372...  135.999178   

                                   Spectral Contrast  \
0  [[ 8.88813143 12.94743439 14.04482159 ...  8.7...   
1  [[17.79094137 11.06542308 11.43209494 ... 19.3...   
2  [[ 9.9482166  17.3706133  26.02688982 ... 20.8...   
3  [[11.78803377 16.61964769 31.87276376 ... 24.0...   
4  [[11.99368033 13.25920714 15.62874838 ... 15.4...   

                                   Spectral Roll-off  \
0  [[4705.00488281 4640.40527344 4715.77148438 .....   
1  [[1625.75683594 2260.98632812 2723.95019531 .....   
2  [[2433.25195312 2368.65234375 2390.18554688 .....   
3  [[1205.859375   1248.92578125 1119.7265625  .....   
4  [[2293.28613281 2788.54980469 3025.41503906 .....   

                                                MFCC  \
0  [[-2.40635422e+02 -2.11214355e+02 -1.93908890e...   
1  [[-3.39146393e+02 -2.23899673e+02 -1.56065292e...   
2  [[-92.80102   -77.73783   -96.38281   ... -84....   
3  [[-2.2065663e+02 -2.0152414e+02 -2.1341223e+02...   
4  [[-281.78323   -180.33914   -155.01845   ... -...   

                                              Chroma  \
0  [[0.33112547 0.2349237  0.8714165  ... 0.15691...   
1  [[0.36823648 0.8158752  1.         ... 0.83394...   
2  [[0.27561098 0.20460801 0.19981515 ... 0.24957...   
3  [[0.60287297 0.36729065 0.33834004 ... 0.55220...   
4  [[0.19081402 0.17394003 0.06381712 ... 0.05633...   

                                   Spectral Centroid  \
0  [[2122.97750086 2102.62107328 1955.28566681 .....   
1  [[ 886.90855361 1051.83134299 1125.26706995 .....   
2  [[1510.21141028 1419.00318588 1386.42492231 .....   
3  [[ 694.20628788  696.3364192   659.96426081 .....   
4  [[1367.54433603 1552.98979661 1630.62768824 .....   

                                  Spectral Bandwidth  \
0  [[2283.0843996  2247.35232089 2308.61530036 .....   
1  [[1322.56608095 1750.98898746 1915.93500528 .....   
2  [[1480.48652053 1525.31517934 1567.39297006 .....   
3  [[1162.25843622 1184.84755668 1146.65869359 .....   
4  [[1544.2611461  1538.50945801 1543.53500673 .....   

                                   Spectral Flatness  \
0  [[0.01456181 0.01049567 0.00578466 ... 0.00196...   
1  [[0.00080758 0.00106355 0.00085136 ... 0.00215...   
2  [[0.00218783 0.00097655 0.00057822 ... 0.00327...   
3  [[1.98361915e-04 1.11818306e-04 6.90561428e-05...   
4  [[0.00295216 0.00145499 0.00069517 ... 0.00108...   

                                      

In [6]:
df['Genre'].unique()

array(['blues', 'classical', 'country', 'disco', 'hiphop', 'jazz',
       'metal', 'pop', 'reggae', 'rock'], dtype=object)

This function loads and plays an audio file of a specific genre and number using the librosa library. It takes two arguments, genre and num, which specify the genre of the audio and the number of the audio file within that genre, respectively.

In [ ]:
df['Beats Mean'] = df['Beats'].apply(lambda x: np.mean(x))
df['Central Moments Mean'] = df['Central Moments'].apply(lambda x: np.mean(x))
df['Zero Crossing Rate Mean'] = df['Zero Crossing Rate'].apply(lambda x: np.mean(x))
df['RMSE Mean'] = df['RMSE'].apply(lambda x: np.mean(x))
df['Spectral Contrast Mean'] = df['Spectral Contrast'].apply(lambda x: np.mean(x))
df['Spectral Roll-off Mean'] = df['Spectral Roll-off'].apply(lambda x: np.mean(x))
df['MFCC Mean'] = df['MFCC'].apply(lambda x: np.mean(x))
df['Chroma Mean'] = df['Chroma'].apply(lambda x: np.mean(x))
df['Spectral Centroid Mean'] = df['Spectral Centroid'].apply(lambda x: np.mean(x))
df['Spectral Bandwidth Mean'] = df['Spectral Bandwidth'].apply(lambda x: np.mean(x))
df['Spectral Flatness Mean'] = df['Spectral Flatness'].apply(lambda x: np.mean(x))
df['Loudness Mean'] = df['Loudness'].apply(lambda x: np.mean(x))
df['Harmonicity Mean'] = df['Harmonicity'].apply(lambda x: np.mean(x))
df['Percussiveness Mean'] = df['Percussiveness'].apply(lambda x: np.mean(x))

In [ ]:
df['Beats std'] = df['Beats'].apply(lambda x: np.std(x))
df['Central Moments std'] = df['Central Moments'].apply(lambda x: np.std(x))
df['Zero Crossing Rate std'] = df['Zero Crossing Rate'].apply(lambda x: np.std(x))
df['RMSE std'] = df['RMSE'].apply(lambda x: np.std(x))
df['Spectral Contrast std'] = df['Spectral Contrast'].apply(lambda x: np.std(x))
df['Spectral Roll-off std'] = df['Spectral Roll-off'].apply(lambda x: np.std(x))
df['MFCC std'] = df['MFCC'].apply(lambda x: np.std(x))
df['Chroma std'] = df['Chroma'].apply(lambda x: np.std(x))
df['Spectral Centroid std'] = df['Spectral Centroid'].apply(lambda x: np.std(x))
df['Spectral Bandwidth std'] = df['Spectral Bandwidth'].apply(lambda x: np.std(x))
df['Spectral Flatness std'] = df['Spectral Flatness'].apply(lambda x: np.std(x))
df['Loudness std'] = df['Loudness'].apply(lambda x: np.std(x))
df['Harmonicity std'] = df['Harmonicity'].apply(lambda x: np.std(x))
df['Percussiveness std'] = df['Percussiveness'].apply(lambda x: np.std(x))

In [ ]:
df['Beats var'] = df['Beats'].apply(lambda x: np.var(x))
df['Central Moments var'] = df['Central Moments'].apply(lambda x: np.var(x))
df['Zero Crossing Rate var'] = df['Zero Crossing Rate'].apply(lambda x: np.var(x))
df['RMSE var'] = df['RMSE'].apply(lambda x: np.var(x))
df['Spectral Contrast var'] = df['Spectral Contrast'].apply(lambda x: np.var(x))
df['Spectral Roll-off var'] = df['Spectral Roll-off'].apply(lambda x: np.var(x))
df['MFCC var'] = df['MFCC'].apply(lambda x: np.var(x))
df['Chroma var'] = df['Chroma'].apply(lambda x: np.var(x))
df['Spectral Centroid var'] = df['Spectral Centroid'].apply(lambda x: np.var(x))
df['Spectral Bandwidth var'] = df['Spectral Bandwidth'].apply(lambda x: np.var(x))
df['Spectral Flatness var'] = df['Spectral Flatness'].apply(lambda x: np.var(x))
df['Loudness var'] = df['Loudness'].apply(lambda x: np.var(x))
df['Harmonicity var'] = df['Harmonicity'].apply(lambda x: np.var(x))
df['Percussiveness var'] = df['Percussiveness'].apply(lambda x: np.var(x))

In [ ]:
df['Beats max'] = df['Beats'].apply(lambda x: np.max(x))
df['Central Moments max'] = df['Central Moments'].apply(lambda x: np.max(x))
df['Zero Crossing Rate max'] = df['Zero Crossing Rate'].apply(lambda x: np.max(x))
df['RMSE max'] = df['RMSE'].apply(lambda x: np.max(x))
df['Spectral Contrast max'] = df['Spectral Contrast'].apply(lambda x: np.max(x))
df['Spectral Roll-off max'] = df['Spectral Roll-off'].apply(lambda x: np.max(x))
df['MFCC max'] = df['MFCC'].apply(lambda x: np.max(x))
df['Chroma max'] = df['Chroma'].apply(lambda x: np.max(x))
df['Spectral Centroid max'] = df['Spectral Centroid'].apply(lambda x: np.max(x))
df['Spectral Bandwidth max'] = df['Spectral Bandwidth'].apply(lambda x: np.max(x))
df['Spectral Flatness max'] = df['Spectral Flatness'].apply(lambda x: np.max(x))
df['Loudness max'] = df['Loudness'].apply(lambda x: np.max(x))
df['Harmonicity max'] = df['Harmonicity'].apply(lambda x: np.max(x))
df['Percussiveness max'] = df['Percussiveness'].apply(lambda x: np.max(x))

In [ ]:
df['Beats min'] = df['Beats'].apply(lambda x: np.min(x))
df['Central Moments min'] = df['Central Moments'].apply(lambda x: np.min(x))
df['Zero Crossing Rate min'] = df['Zero Crossing Rate'].apply(lambda x: np.min(x))
df['RMSE min'] = df['RMSE'].apply(lambda x: np.min(x))
df['Spectral Contrast min'] = df['Spectral Contrast'].apply(lambda x: np.min(x))
df['Spectral Roll-off min'] = df['Spectral Roll-off'].apply(lambda x: np.min(x))
df['MFCC min'] = df['MFCC'].apply(lambda x: np.min(x))
df['Chroma min'] = df['Chroma'].apply(lambda x: np.min(x))
df['Spectral Centroid min'] = df['Spectral Centroid'].apply(lambda x: np.min(x))
df['Spectral Bandwidth min'] = df['Spectral Bandwidth'].apply(lambda x: np.min(x))
df['Spectral Flatness min'] = df['Spectral Flatness'].apply(lambda x: np.min(x))
df['Loudness min'] = df['Loudness'].apply(lambda x: np.min(x))
df['Harmonicity min'] = df['Harmonicity'].apply(lambda x: np.min(x))
df['Percussiveness min'] = df['Percussiveness'].apply(lambda x: np.min(x))

In [ ]:
df['Beats sum'] = df['Beats'].apply(lambda x: np.sum(x))
df['Central Moments sum'] = df['Central Moments'].apply(lambda x: np.sum(x))
df['Zero Crossing Rate sum'] = df['Zero Crossing Rate'].apply(lambda x: np.sum(x))
df['RMSE sum'] = df['RMSE'].apply(lambda x: np.sum(x))
df['Spectral Contrast sum'] = df['Spectral Contrast'].apply(lambda x: np.sum(x))
df['Spectral Roll-off sum'] = df['Spectral Roll-off'].apply(lambda x: np.sum(x))
df['MFCC sum'] = df['MFCC'].apply(lambda x: np.sum(x))
df['Chroma sum'] = df['Chroma'].apply(lambda x: np.sum(x))
df['Spectral Centroid sum'] = df['Spectral Centroid'].apply(lambda x: np.sum(x))
df['Spectral Bandwidth sum'] = df['Spectral Bandwidth'].apply(lambda x: np.sum(x))
df['Spectral Flatness sum'] = df['Spectral Flatness'].apply(lambda x: np.sum(x))
df['Loudness sum'] = df['Loudness'].apply(lambda x: np.sum(x))
df['Harmonicity sum'] = df['Harmonicity'].apply(lambda x: np.sum(x))
df['Percussiveness sum'] = df['Percussiveness'].apply(lambda x: np.sum(x))

In [ ]:
df['Beats_Above_Mean'] = df['Beats'].apply(count_values_above_mean)
df['Central Moments_Above_Mean'] = df['Central Moments'].apply(count_values_above_mean)
df['Zero Crossing Rate_Above_Mean'] = df['Zero Crossing Rate'].apply(count_values_above_mean)
df['RMSE_Above_Mean'] = df['RMSE'].apply(count_values_above_mean)
df['Spectral Contrast_Above_Mean'] = df['Spectral Contrast'].apply(count_values_above_mean)
df['Spectral Roll-off_Above_Mean'] = df['Spectral Roll-off'].apply(count_values_above_mean)
df['MFCC_Above_Mean'] = df['MFCC'].apply(count_values_above_mean)
df['Chroma_Above_Mean'] = df['Chroma'].apply(count_values_above_mean)
df['Spectral Centroid_Above_Mean'] = df['Spectral Centroid'].apply(count_values_above_mean)
df['Spectral Bandwidth_Above_Mean'] = df['Spectral Bandwidth'].apply(count_values_above_mean)
df['Spectral Flatness_Above_Mean'] = df['Spectral Flatness'].apply(count_values_above_mean)
df['Loudness_Above_Mean'] = df['Loudness'].apply(count_values_above_mean)
df['Harmonicity_Above_Mean'] = df['Harmonicity'].apply(count_values_above_mean)
df['Percussiveness_Above_Mean'] = df['Percussiveness'].apply(count_values_above_mean)

<a id='data-visualization'></a>
<font size="+1" color='#780404'><b> 4.2 Data Visualization</b></font>  
[back to top](#table-of-contents)

This code generates a frequency bar chart of the 'Tempo' column in a pandas DataFrame.

In [13]:
if isinstance(df, (pd.DatetimeIndex, pd.MultiIndex)):
    df = df.to_frame(index=False)
df = df.reset_index().drop('index', axis=1, errors='ignore')
df.columns = [str(c) for c in df.columns]
tempo_counts = df['Tempo'].value_counts().reset_index()
tempo_counts.columns = ['Tempo', 'Frequency']
tempo_counts = tempo_counts.sort_values(['Frequency', 'Tempo'], ascending=[False, True])
tempo_counts = tempo_counts[:100]
sns.set(style="whitegrid")
plt.figure(figsize=(10, 8))
ax = sns.barplot(x='Frequency', y='Tempo', data=tempo_counts)
ax.set(xlabel='Frequency', ylabel='Tempo', title='Tempo Value Counts')
plt.show()


In [14]:
def show_waveform(x, num):
    # Load WAV file
    wav_file = f'../data/genres/{x}/{x}.{num}.wav'
    y, sr = librosa.load(wav_file)
# Create x-axis values
    time = librosa.times_like(y, sr=sr)

    sns.set(style="whitegrid")
    plt.figure(figsize=(10, 8))
    sns.lineplot(x=time, y=y)
    plt.xlabel('Time (s)')
    plt.ylabel('Amplitude')
    plt.title(f'Sample waveform for {x}')
    plt.show()


This function loads a WAV file of a specific genre and number, and generates a sample waveform plot using the librosa and plotly libraries. It takes two arguments, x and num, which specify the genre of the audio and the number of the audio file within that genre, respectively.

In [11]:
show_waveform('blues', '00090')

In [ ]:
show_waveform('classical', '00090')

In [ ]:
show_waveform('country', '00090')

In [ ]:
show_waveform('disco', '00090')

In [ ]:
show_waveform('hiphop', '00090')

In [ ]:
show_waveform('jazz', '00090')

In [ ]:
show_waveform('metal', '00090')

In [ ]:
show_waveform('pop', '00090')

In [ ]:
show_waveform('reggae', '00090')

In [ ]:
show_waveform('rock', '00090')

In [ ]:
def show_spectogram(x, num):
    # Load audio file
    audio_path = f'/Users/umarkabir/Documents/Qwasar/Classical Punk/genres/{x}/{x}.{num}.wav'
    y, sr = librosa.load(audio_path)

    # Calculate spectrogram
    S = librosa.feature.melspectrogram(y=y, sr=sr, n_mels=128, fmax=8000)

    # Convert to decibels
    S_dB = librosa.power_to_db(S, ref=np.max)

    # Create figure
    plt.figure(figsize=(10, 4))
    ax = sns.heatmap(S_dB, cmap='viridis')

    # Set x and y axis labels
    ax.set_xlabel('Time')
    ax.set_ylabel('Frequency (Hz)')

    # Set figure title
    ax.set_title(f'Sample spectrogram for {x}')

    # Show figure
    plt.show()

This function loads a WAV file of a specific genre and number, and generates a sample spectrogram plot using the librosa and plotly libraries. It takes two arguments, x and num, which specify the genre of the audio and the number of the audio file within that genre, respectively.

In [ ]:
show_spectogram('blues', '00090')

In [ ]:
show_spectogram('classical', '00090')

In [ ]:
show_spectogram('country', '00090')

In [ ]:
show_spectogram('disco', '00090')

In [ ]:
show_spectogram('hiphop', '00090')

In [ ]:
show_spectogram('jazz', '00090')

In [ ]:
show_spectogram('metal', '00090')

In [ ]:
show_spectogram('pop', '00090')

In [ ]:
show_spectogram('reggae', '00090')

In [ ]:
show_spectogram('rock', '00090')

In [ ]:
import seaborn as sns

def show_sr(x, num):
    # Load audio file
    audio_path = f'/Users/umarkabir/Documents/Qwasar/Classical Punk/genres/{x}/{x}.{num}.wav'
    y, sr = librosa.load(audio_path)

    # Compute spectral rolloff
    spectral_rolloff = librosa.feature.spectral_rolloff(y=y, sr=sr)[0]

    # Create plot
    sns.set(style="whitegrid")
    fig, ax = plt.subplots(figsize=(10, 6))
    ax.plot(spectral_rolloff, color='blue')
    ax.set(title=f'Sample spectral rolloff for {x}', xlabel='Frame', ylabel='Frequency (Hz)')
    plt.show()


This function show_sr(x, num) loads an audio file and computes the spectral rolloff. It then creates a Plotly line plot of the spectral rolloff values with the x-axis representing the frame and the y-axis representing frequency in Hz. The title of the plot is set to "Sample spectral rolloff for x", where x is the name of the audio file.

In [ ]:
show_sr('blues', '00090')

In [ ]:
show_sr('classical', '00090')

In [ ]:
show_sr('country', '00090')

In [ ]:
show_sr('disco', '00090')

In [ ]:
show_sr('hiphop', '00090')

In [ ]:
show_sr('jazz', '00090')

In [ ]:
show_sr('metal', '00090')

In [ ]:
show_sr('pop', '00090')

In [ ]:
show_sr('reggae', '00090')

In [ ]:
show_sr('rock', '00090')

In [ ]:
def show_chroma(x, num):
    # Load audio file
    audio_path = f'/Users/umarkabir/Documents/Qwasar/Classical Punk/genres/{x}/{x}.{num}.wav'
    y, sr = librosa.load(audio_path)

    # Compute chroma feature
    chroma = librosa.feature.chroma_stft(y=y, sr=sr)

    # Create time axis in seconds
    time = librosa.frames_to_time(np.arange(chroma.shape[1]), sr=sr)

    # Create chroma note names
    chroma_note_names = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B']

    # Create dataframe
    df = pd.DataFrame(chroma, index=chroma_note_names, columns=time)

    # Plot heatmap using seaborn
    plt.figure(figsize=(10, 8))
    sns.heatmap(df, cmap='viridis', xticklabels=50, yticklabels=1)
    plt.title(f'Sample chroma feature for {x}')
    plt.xlabel('Time (s)')
    plt.ylabel('Chroma Note')
    plt.show()

This function takes in two arguments x and num representing the music genre and the song number, respectively. It then loads the corresponding audio file and computes the chroma feature using librosa's chroma_stft function. It creates a time axis in seconds using frames_to_time function, and chroma note names as a list. It then creates a heatmap trace using go.Heatmap with time as the x-axis, chroma_note_names as the y-axis, and chroma as the z-axis. Finally, it sets the layout with appropriate x and y axis titles, and a title for the figure. It shows the resulting figure using fig.show().

In [ ]:
show_chroma('blues', '00090')

In [ ]:
show_chroma('classical', '00090')

In [ ]:
show_chroma('country', '00090')

In [ ]:
show_chroma('disco', '00090')

In [ ]:
show_chroma('hiphop', '00090')

In [ ]:
show_chroma('jazz', '00090')

In [ ]:
show_chroma('metal', '00090')

In [ ]:
show_chroma('pop', '00090')

In [ ]:
show_chroma('reggae', '00090')

In [ ]:
show_chroma('rock', '00090')

In [ ]:
def show_zcr(x, num):
    # Load audio file
    audio_path = f'/Users/umarkabir/Documents/Qwasar/Classical Punk/genres/{x}/{x}.{num}.wav'
    y, sr = librosa.load(audio_path)

    # Compute zero crossing rate
    zcr = librosa.feature.zero_crossing_rate(y)

    # Plot with Seaborn
    fig, ax = plt.subplots(figsize=(10,6))
    sns.lineplot(data=zcr[0], ax=ax)
    ax.set(title=f'Zero Crossing Rate for {x} Genre', xlabel='Time (s)', ylabel='ZCR')
    plt.show()


The show_zcr function takes in two arguments: x, which represents the genre of the music file, and num, which represents the number of the music file. It calculates the zero-crossing rate of the audio file and plots it using Plotly. 

In [ ]:
show_zcr('blues', '00090')

In [ ]:
show_zcr('classical', '00090')

In [ ]:
show_zcr('country', '00090')

In [ ]:
show_zcr('disco', '00090')

In [ ]:
show_zcr('hiphop', '00090')

In [ ]:
show_zcr('jazz', '00090')

In [ ]:
show_zcr('pop', '00090')

In [ ]:
show_zcr('reggae', '00090')

In [ ]:
show_zcr('rock', '00090')

<a id='summary-statistics'></a>
<font size="+1" color='#780404'><b> 4.3 Summary Statistics</b></font>  
[back to top](#table-of-contents)

In [ ]:
df.describe(include='all')

In [ ]:
def skew_kurt(data, col):
    # Calculate skewness and kurtosis of Income column
    _skewness = skew(data[col])
    _kurtosis = kurtosis(data[col])

    # Create histogram of Income column with mean, median, and mode
    sns.histplot(data=data, x=col, kde=True)
    plt.axvline(data[col].mean(), color='r', linestyle='--', label='Mean')
    plt.axvline(data[col].median(), color='g', linestyle='--', label='Median')
    plt.axvline(data[col].mode()[0], color='b', linestyle='--', label='Mode')
    plt.legend()

    # Add text annotation for skewness and kurtosis values
    plt.annotate('Skewness: {:.2f}'.format(_skewness), xy=(0.5, 0.9), xycoords='axes fraction')
    plt.annotate('Kurtosis: {:.2f}'.format(_kurtosis), xy=(0.5, 0.85), xycoords='axes fraction')

    plt.show()

In [ ]:
skew_kurt(df[df['Genre'] == 'blues'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'classical'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'country'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'disco'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'hiphop'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'jazz'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'metal'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'pop'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'reggae'], 'Tempo')

In [ ]:
skew_kurt(df[df['Genre'] == 'rock'], 'Tempo')

<a id='feature-correlation'></a>
<font size="+1" color='#780404'><b> 4.4 Feature Correlation</b></font>  
[back to top](#table-of-contents)

In [ ]:
df_corr = df.corr()

In [ ]:
# Compute correlation matrix

# Set figure size and font sizes
fig, ax = plt.subplots(figsize=(50, 50))
sns.set(font_scale=1.9)

# Plot heatmap with adjusted color map
sns.heatmap(df_corr, cmap='coolwarm', annot=True, center=0, square=True)

# Adjust font size of features
ax.set_xticklabels(ax.get_xticklabels(), fontsize=35)
ax.set_yticklabels(ax.get_yticklabels(), fontsize=35)

# Add title and axis labels
plt.title('Correlation Matrix', fontsize=30)
plt.xlabel('Features', fontsize=20)
plt.ylabel('Features', fontsize=20)

# Show plot
plt.show()

<a id='data-preparation'></a>
<font size="+2" color='#053c96'><b> 5. Data Preparation</b></font>  
[back to top](#table-of-contents)

<a id='data-cleaning'></a>
<font size="+1" color='#780404'><b> 5.1 Data Cleaning</b></font>  
[back to top](#table-of-contents)

<a id='feature-engineering'></a>
<font size="+1" color='#780404'><b> 5.2 Feature Engineering</b></font>  
[back to top](#table-of-contents)

In [ ]:
df['Beats Mean'] = df['Beats'].apply(lambda x: np.mean(x))
df['Central Moments Mean'] = df['Central Moments'].apply(lambda x: np.mean(x))
df['Zero Crossing Rate Mean'] = df['Zero Crossing Rate'].apply(lambda x: np.mean(x))
df['RMSE Mean'] = df['RMSE'].apply(lambda x: np.mean(x))
df['Spectral Contrast Mean'] = df['Spectral Contrast'].apply(lambda x: np.mean(x))
df['Spectral Roll-off Mean'] = df['Spectral Roll-off'].apply(lambda x: np.mean(x))
df['MFCC Mean'] = df['MFCC'].apply(lambda x: np.mean(x))
df['Chroma Mean'] = df['Chroma'].apply(lambda x: np.mean(x))
df['Spectral Centroid Mean'] = df['Spectral Centroid'].apply(lambda x: np.mean(x))
df['Spectral Bandwidth Mean'] = df['Spectral Bandwidth'].apply(lambda x: np.mean(x))

In [ ]:
df['Beats Var'] = df['Beats'].apply(lambda x: np.var(x))
df['Central Moments Var'] = df['Central Moments'].apply(lambda x: np.var(x))
df['Zero Crossing Rate Var'] = df['Zero Crossing Rate'].apply(lambda x: np.var(x))
df['RMSE Var'] = df['RMSE'].apply(lambda x: np.var(x))
df['Spectral Contrast Var'] = df['Spectral Contrast'].apply(lambda x: np.var(x))
df['Spectral Roll-off Var'] = df['Spectral Roll-off'].apply(lambda x: np.var(x))
df['MFCC Var'] = df['MFCC'].apply(lambda x: np.var(x))
df['Chroma Var'] = df['Chroma'].apply(lambda x: np.var(x))
df['Spectral Centroid Var'] = df['Spectral Centroid'].apply(lambda x: np.var(x))
df['Spectral Bandwidth Var'] = df['Spectral Bandwidth'].apply(lambda x: np.var(x))

In [ ]:
df['Beats Std'] = df['Beats'].apply(lambda x: np.std(x))
df['Central Moments Std'] = df['Central Moments'].apply(lambda x: np.std(x))
df['Zero Crossing Rate Std'] = df['Zero Crossing Rate'].apply(lambda x: np.std(x))
df['RMSE Std'] = df['RMSE'].apply(lambda x: np.std(x))
df['Spectral Contrast Std'] = df['Spectral Contrast'].apply(lambda x: np.std(x))
df['Spectral Roll-off Std'] = df['Spectral Roll-off'].apply(lambda x: np.std(x))
df['MFCC Std'] = df['MFCC'].apply(lambda x: np.std(x))
df['Chroma Std'] = df['Chroma'].apply(lambda x: np.std(x))
df['Spectral Centroid Std'] = df['Spectral Centroid'].apply(lambda x: np.std(x))
df['Spectral Bandwidth Std'] = df['Spectral Bandwidth'].apply(lambda x: np.std(x))

<a id='data-transformation'></a>
<font size="+1" color='#780404'><b> 5.3 Data Transformation</b></font>  
[back to top](#table-of-contents)
<a id='modeling'></a>

In [ ]:
genre_col = df['Genre']
le = LabelEncoder()
y = le.fit_transform(genre_col)

In [ ]:
# create target dataset
#y = df['Genre']

In [ ]:
x = df.drop(['Genre', 'Filename', 'Unnamed: 0', 'Beats', 'SR',
             'Central Moments', 'Zero Crossing Rate', 'RMSE',
             'Spectral Contrast', 'Spectral Roll-off', 'MFCC',
             'Chroma', 'Spectral Centroid', 'Spectral Bandwidth'], axis=1)

In [ ]:
x.shape

In [ ]:
#Create a TF-IDF object instance
scaler = StandardScaler()
X = scaler.fit_transform(x)

In [ ]:
#Pareto Principle Split
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size= 0.3,
                                                    random_state= 42)

<a id='modeling'></a>

<font size="+2" color='#053c96'><b> 6. Modeling</b></font>  
[back to top](#table-of-contents)

<a id='model-selection'></a>

<font size="+1" color='#780404'><b> 6.1 Model Selection</b></font>  
[back to top](#table-of-contents)

In [ ]:
# Multiclass Support Vector Machines (SVM)
svm_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', OneVsRestClassifier(SVC(kernel='rbf', C=1, gamma='scale')))
])

# K-Means Clustering
kmeans_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('kmeans', KMeans(n_clusters=10))
])

# K-Nearest Neighbors (KNN)
knn_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('knn', KNeighborsClassifier(n_neighbors=5))
])

# Convolutional Neural Networks (CNNs)
cnn_pipeline = Pipeline([
    ('clf', Sequential([
        Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        MaxPooling2D((2, 2)),
        Flatten(),
        Dense(10, activation='softmax')
    ]))
])

# Compile the CNN model
cnn_pipeline.named_steps['clf'].compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


<a id='model-training'></a>

<font size="+1" color='#780404'><b> 6.2 Model Training</b></font>  
[back to top](#table-of-contents)

In [ ]:
# Fit SVM pipeline
svm_pipeline.fit(X_train, y_train)

# Fit K-Means pipeline
kmeans_pipeline.fit(X_train)

# Fit KNN pipeline
knn_pipeline.fit(X_train, y_train)


<a id='model-evaluation'></a>

<font size="+1" color='#780404'><b> 6.3 Model Evaluation</b></font>  
[back to top](#table-of-contents)

In [ ]:
# Initialize the lists for performance metrics and confusion matrices
f1_list = []
pr_list = []
rc_list = []
acc_list = []
cm_list = []

# Evaluate additional models
# evaluate the SVM pipeline

y_pred_svm = svm_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_svm, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_svm, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_svm, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_svm))
cm_list.append(confusion_matrix(y_test, y_pred_svm))

# evaluate the K-Means pipeline
y_pred_kmeans = kmeans_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_kmeans, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_kmeans, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_kmeans, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_kmeans))
cm_list.append(confusion_matrix(y_test, y_pred_kmeans))

# evaluate the KNN pipeline
y_pred_knn = knn_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_knn, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_knn, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_knn, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_knn))
cm_list.append(confusion_matrix(y_test, y_pred_knn))

# evaluate the CNN pipeline
y_pred_cnn = cnn_pipeline.predict(X_test)
f1_list.append(f1_score(y_test, y_pred_cnn, average='weighted'))
pr_list.append(precision_score(y_test, y_pred_cnn, average='weighted'))
rc_list.append(recall_score(y_test, y_pred_cnn, average='weighted'))
acc_list.append(accuracy_score(y_test, y_pred_cnn))
cm_list.append(confusion_matrix(y_test, y_pred_cnn))

In [ ]:
model_list = ['SVM', 'Kmeans', 'KNN']

<a id='results'></a>
<font size="+2" color='#053c96'><b> 7. Results</b></font>  
[back to top](#table-of-contents)

In [ ]:
result = pd.DataFrame({'Model': model_list, 'F1': f1_list, 'Precision': pr_list, 'Recall': rc_list, 'Accuracy': acc_list})

In [ ]:
result

<a id='analysis-results'></a>

<font size="+1" color='#780404'><b> 7.1 Analysis of Results</b></font>  
[back to top](#table-of-contents)  

The results suggest that the SVM and KNN models perform better than the Kmeans model, based on F1-score, precision, recall, and accuracy.  

Overall, the SVM and KNN models appear to be more suitable for this task than the Kmeans model. However, the performance of the models may depend on the specific characteristics of the dataset, and further analysis may be necessary to confirm the generalizability of the models.

<a id='model-performance'></a>

<font size="+1" color='#780404'><b> 7.2 Model Performance</b></font>  
[back to top](#table-of-contents)  

From the given results, the SVM and KNN models perform better than the Kmeans model in terms of F1-score, precision, and recall. The SVM model has the highest F1-score and recall, while the KNN model has the highest precision. However, the SVM model has slightly lower precision than the KNN model.

The Kmeans model has the lowest F1-score, precision, and recall among the three models, indicating that it performs poorly in predicting positive cases. Although the accuracy of the Kmeans model is the same as that of the SVM model, accuracy alone is not a reliable metric for imbalanced datasets where the majority class dominates the prediction.

Therefore, based on the given results, the SVM and KNN models appear to be more suitable for the given task than the Kmeans model. However, it is important to note that the performance of the models may depend on the specific characteristics of the dataset, and further analysis may be necessary to confirm the generalizability of the models.

<a id='implications'></a>

<font size="+1" color='#780404'><b> 7.3 Implications</b></font>  
[back to top](#table-of-contents)  

The given results have several implications, such as:

Model Selection: The results suggest that the SVM and KNN models may be more suitable for this task than the Kmeans model. Therefore, the SVM or KNN model may be selected for further analysis or implementation.

Hyperparameter Tuning: The performance of the models may depend on the hyperparameters chosen. Therefore, hyperparameter tuning can be performed to optimize the model performance.

Dataset Characteristics: The performance of the models may depend on the specific characteristics of the dataset, such as the distribution of the classes or the presence of outliers. Therefore, further analysis may be necessary to confirm the generalizability of the models to other datasets.

Performance Metrics: The selection of performance metrics can have implications for the evaluation of the models. For example, the accuracy metric may not be reliable for imbalanced datasets. Therefore, multiple metrics should be used to evaluate the models comprehensively.

Overall, the given results provide insights into the performance of the models and can guide further analysis or implementation of the models.

<a id='conclusion'></a>

<font size="+2" color='#053c96'><b> 8. Conclusion</b></font>  
[back to top](#table-of-contents)

<a id='summary'></a>

<font size="+1" color='#780404'><b> 8.1 Summary</b></font>  
[back to top](#table-of-contents)  

In conclusion, the results of the machine learning models developed for music genre classification suggest that the SVM and KNN models perform better than the Kmeans model. Both the SVM and KNN models have higher F1-scores, precision, and recall than the Kmeans model, indicating better performance in predicting the different music genres, including Classically Punk. This implies that the developed machine learning application can automate the process of music genre classification, making it faster, more accurate, and scalable. The application can extract features from audio files and train an SVM or KNN model to classify music into different genres without the need for human intervention. However, further analysis is necessary to confirm the generalizability of the models to other datasets, and hyperparameter tuning may be required to optimize the performance of the models. Overall, the developed machine learning application has the potential to revolutionize the music industry by streamlining the process of music genre classification.

<a id='recommendations'></a>

<font size="+1" color='#780404'><b> 8.2 Recommendations</b></font>  
[back to top](#table-of-contents)  

Based on the problem statement and the results obtained from the machine learning models, here are some recommendations that could help to further improve the music genre classification application:

1. Data Augmentation: To improve the performance of the machine learning models, it may be helpful to augment the training data by adding variations of the existing data. This can be done by applying audio effects such as pitch shifting, time stretching, or adding noise to create a more diverse training set.

2. Feature Engineering: Feature engineering involves selecting and extracting meaningful features from audio files to train the machine learning model. By experimenting with different audio features and selecting the most informative ones, the performance of the machine learning model can be improved.

3. Ensemble Learning: Ensemble learning involves combining the predictions of multiple machine learning models to improve the overall performance. By training multiple models using different algorithms or hyperparameters and combining their predictions, the overall accuracy of the application can be improved.

4. User Feedback: Collecting feedback from users of the application can help to improve the accuracy of the machine learning model over time. By allowing users to provide feedback on the accuracy of the classification results, the application can learn from its mistakes and continuously improve its performance.

5. Continuous Monitoring: It is important to continuously monitor the performance of the machine learning model to ensure that it is accurately classifying the different music genres. Regularly updating the model with new data and retraining it can help to improve its accuracy and keep up with changes in the music industry.

<a id='references'></a>

<font size="+2" color='#053c96'><b> 9. References</b></font>  
[back to top](#table-of-contents)  

1. Arsh Chowdhry. Music Genre Classification Using CNN. https://blog.clairvoyantsoft.com/music-genre-classification-using-cnn-ef9461553726
2. Faisal Ahmed, Padma Polash Paul, Marina Gavrilova. Music Genre Classification Using a Gradient-Based Local Texture Descriptor https://www.researchgate.net/publication/303860385_Music_Genre_Classification_Using_a_Gradient-Based_Local_Texture_Descriptor
3. Albert Jiménez. Music Genre Classification with Deep Learning https://github.com/jsalbert/Music-Genre-Classification-with-Deep-Learning
4. Insiyah Hajoori. Music Genre Classification https://github.com/Insiyaa/Music-Genre-Classification
5. A. Elbir, N. Aydin. Music genre classification and music recommendation by using deep learning https://ietresearch.onlinelibrary.wiley.com/doi/full/10.1049/el.2019.4202